src/networking.c

```diff
@@ -4317,44 +4318,70 @@ void killIOThreads(void) {
     }
 }
 
-void startThreadedIO(void) {
-    serverAssert(server.io_threads_active == 0);
-    for (int j = 1; j < server.io_threads_num; j++)
-        pthread_mutex_unlock(&io_threads_mutex[j]);
-    server.io_threads_active = 1;
+/* This function will active additional num I/O threads. */
+static inline void activateThreadedIO(int num) {
+    serverAssert(num > 0);
+    serverAssert(server.io_threads_active_num + num <= server.io_threads_num);
+    for (int i = 0; i < num; i++)
+        pthread_mutex_unlock(&io_threads_mutex[server.io_threads_active_num + i]);
+    server.io_threads_active_num += num;
 }
 
-void stopThreadedIO(void) {
+/* This function will inactive additional num I/O threads. */
+static inline void deactivateThreadedIO(int num) {
+    serverAssert(num > 0);
+    serverAssert(server.io_threads_active_num - num >= 1);
     /* We may have still clients with pending reads when this function
      * is called: handle them before stopping the threads. */
     handleClientsWithPendingReadsUsingThreads();
-    serverAssert(server.io_threads_active == 1);
-    for (int j = 1; j < server.io_threads_num; j++)
-        pthread_mutex_lock(&io_threads_mutex[j]);
-    server.io_threads_active = 0;
+    for (int i = 0; i < num; i++)
+        pthread_mutex_lock(&io_threads_mutex[server.io_threads_active_num -1 - i]);
+    server.io_threads_active_num -= num;
 }
 
-/* This function checks if there are not enough pending clients to justify
- * taking the I/O threads active: in that case I/O threads are stopped if
- * currently active. We track the pending writes as a measure of clients
- * we need to handle in parallel, however the I/O threading is disabled
- * globally for reads as well if we have too little pending clients.
+/* This function will dynamically adjust the I/O threads number according to
+ * real workloads in runtime, currently we track the latest x rounds of
+ * pending writes as a measure of clients we need to handle in parallel,
+ * however the I/O threading is disabled globally for reads as well if
+ * we have too little pending clients.
  *
  * The function returns 0 if the I/O threading should be used because there
  * are enough active threads, otherwise 1 is returned and the I/O threads
  * could be possibly stopped (if already active) as a side effect. */
-int stopThreadedIOIfNeeded(void) {
-    int pending = listLength(server.clients_pending_write);
-
-    /* Return ASAP if IO threads are disabled (single threaded mode). */
+#define IO_THREAD_BATCH_NUM 2
+int adjustThreadedIOIfNeeded(void) {
```

Comment:
With the function name change, the return value (of 0 enabling IO threading and 1 disabling IO threading) is confusing. I think the old name is fine (and clearer).

Reply:
Revert name from `adjustThreadedIOIfNeeded` to `stopThreadedIOIfNeeded`